src/clj/fluree/db/reasoner.cljc

```diff
@@ -0,0 +1,276 @@
+(ns fluree.db.reasoner
+  (:require [clojure.core.async :as async :refer [alts! go]]
+            [clojure.string :as str]
+            [fluree.db.flake :as flake]
+            [fluree.db.query.fql.parse :as parse]
+            [fluree.db.util.core :as util :refer [try* catch*]]
+            [fluree.db.util.log :as log]
+            [fluree.db.util.async :refer [go-try <?]]
+            [fluree.db.reasoner.resolve :as resolve]
+            [fluree.db.json-ld.transact :as transact]
+            [fluree.db.fuel :as fuel]
+            [fluree.db.constants :as const]
+            [fluree.json-ld :as json-ld]
+            [fluree.db.query.fql.parse :as q-parse]
+            [fluree.db.reasoner.owl-datalog :as owl-datalog]
+            [fluree.db.dbproto :refer [db?]]
+            [fluree.db.reasoner.graph :refer [task-queue add-rule-dependencies]]))
+
+#?(:clj (set! *warn-on-reflection* true))
+
+(defn reasoned-rule
+  "Returns truthy if the flake has been generated by reasoner"
+  [flake]
+  (-> flake meta :reasoned))
+
+(defn non-reasoned-flakes
+  "Takes a sequence of flakes and removes any flakes which are reasoned.
+
+  This is primarily used to remove reasoned flakes from commits."
+  [flakes]
+  (remove reasoned-rule flakes))
+
+(defn reasoned-flakes
+  "Takes a sequence of flakes and keeps only reasoned flakes"
+  [flakes]
+  (filter reasoned-rule flakes))
+
+(defn schedule
+  "Returns list of rule @id values in the order they should be run.
+
+  If optional result-summary is provided, rules that don't need to be
+  re-run will be filtered out.
+
+  A result summary is list/set of the rule dependency patterns which
+  match newly created Flakes from the last run. When the result-summary
+  is empty, no rules will be left to run, but based on the dependencies
+  it is possible no rules will be left to run even if the result-summary
+  is non-empty"
+  ([rules]
+   (task-queue rules))
+  ([rules result-summary]
+   (task-queue rules result-summary)))
+
+(defn reasoner-insert
+  "When triples from rules require explicit inserts, returns flakes."
+  [db fuel-tracker txn-id author-did rule-id insert-smt]
+  (go-try
+    (let [tx-state      (-> (transact/->tx-state db txn-id author-did rule-id)
+                            (assoc :stage-update? true))
+          flakes-ch     (transact/generate-flakes db fuel-tracker insert-smt tx-state)
+          fuel-error-ch (:error-ch fuel-tracker)
+          chans         (remove nil? [fuel-error-ch flakes-ch])
+          [flakes] (alts! chans :priority true)]
+      (if (util/exception? flakes)
+        (throw flakes)
+        (let [[db* new-flakes] flakes
+              tx-state* (assoc tx-state :stage-update? true)
+              {:keys [db-after add]} (transact/final-db db* new-flakes tx-state*)]
```

Comment:
This step bypasses post-stage db validation. Could there be a situation where the addition of reasoned flakes breaks existing shacl rules or policy?

Reply:
Absolutely there can be, which I'd consider an issue if the data were to be persisted in a commit. In this case the reasoned data is virtual only for queries... so SHACL and transaction policy are both commit-related.

While not there, I see a good option as running query policy post-reasoning - meaning you can reason "virtual" facts that can be used as part of policy rules, or have policy rules that target these virtual facts.